src/main/java/net/glowstone/block/blocktype/BlockCauldron.java

```diff
@@ -36,11 +38,23 @@ public boolean blockInteract(GlowPlayer player, GlowBlock block, BlockFace face,
             return super.blockInteract(player, block, face, clickedLoc);
         }
 
+        // if player is holding a "potion" and not an "empty bottle"
+        if (player.getItemInHand().getType() == Material.POTION
+                && new GlowMetaPotion(player.getItemInHand().getItemMeta()).getBasePotionData().getType() != PotionType.WATER) {
+            return super.blockInteract(player, block, face, clickedLoc);
+        }
+
         switch (player.getItemInHand().getType()) {
-            // todo: filling the cauldron by emptying a bottle
-            // todo: emptying the cauldron by filling a bucket
             case WATER_BUCKET:
-                fillCauldron(player, block);
+                emptyBucket(player, block);
+                return true;
+
+            case BUCKET:
+                fillBucket(player, block);
+                return true;
+
+            case POTION:
+                emptyBottle(player, block);
```

Comment:
Change emptyBottle to return boolean then change this to

```java
if (emptyBottle(player, block) {
  return true;
}
break;
...
} // switch statement end
return super.blockInteract(player, block, face, clickedLoc);
```

Return false if PotionType != WATER.

Reply:
[Line 44](https://github.com/GlowstoneMC/Glowstone/pull/1097/files/c317b67b28b66bd0b06040f30cfe305cdb97a01b#diff-b80570db289b0cfacd9df3c246576452R44) already does this. Am I missing something?

Reply:
I was suggesting an alternative to that logic, which is more extensible for the future if we have more edge case checks we need to do for each item type.

Reply:
Okay, makes sense.